File: binary_trees/preorder_traversal/preorder_traversal.js
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    };
}

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

//      a
//     / \
//    b   c
//   /\   /
//  d  e f
a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.left = f;

function preOrderTraversal(root) {
    let results = [];
    
    function recursion(root) {
        if(!root) return null;
        results.push(root.val);
        recursion(root.left);
        recursion(root.right);
    }

    recursion(root);

    return results;
}

console.log(preOrderTraversal(a))

File: binary_trees/preorder_traversal/preorder_traversal.py
from typing import Optional
class Node:
    def __init__(self, val):
        self.val = val
        self. left = None
        self.right = None

    def __str__(self):
        return f"value: {self.val}, left: {self.left}, right: {self.right}"

#      a
#     / \
#    b   c
#   /\   /
#  d  e f
a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.left = f

def preorder_traversal_recursion(root: Optional[Node]):
    results = []

    def recursion(root):
        if root == None:
            return None
        results.append(root.val)
        recursion(root.left)
        recursion(root.right)

    recursion(root)
    
    return results

print(preorder_traversal_recursion(a))
